### PR TITLE
SUBMARINE-1000. Add submarine save model method

### PR DESCRIPTION
**Pull request by jeff-901**
### What is this PR for?
Create submarine save model method.

### What type of PR is it?
Feature

### Todos
* [x] - Create repository
* [x] - Create save method for pytorch and tensorflow
* [ ] - Create model version when saving.

### What is the Jira issue?
https://issues.apache.org/jira/browse/SUBMARINE-1000

### How should this be tested?
Call the save_model_submarine function and check the artifact by minio ui.
### Screenshots (if appropriate)
Screenshot from 2021-08-31 14-57-59

### Questions:
* Do the license files need updating? No
* Are there breaking changes for older versions? No
* Does this need new documentation? Yes

### CODE REVIEW

**Comment by jeff-901** [comment]
@KUAN-HSUN-LI helps me review the code please.

File: submarine-sdk/pysubmarine/submarine/models/client.py [hunk 1]
```diff
@@ -101,6 +107,39 @@ def save_model(self,
...
+            submarine.models.pytorch.save_model(model, local_dir)
+        elif model_type == "tensorflow":
+            import submarine.models.tensorflow
+
+            submarine.models.tensorflow.save_model(model, local_dir)
+        else:
+            raise Exception(
+                "No valid type of model has been matched to {}".format(
+                    model_type))
+        self.artifact_repo.log_artifacts(local_dir, artifact_path)
```
Comment (KUAN-HSUN-LI) [inline submarine-sdk/pysubmarine/submarine/models/client.py:140]: require to delete the local model

**Review by KUAN-HSUN-LI — COMMENTED** [review]
@jeff-901 There are several suggestions. Thanks

File: submarine-sdk/pysubmarine/submarine/models/client.py [hunk 2]
```diff
@@ -101,6 +107,39 @@ def save_model(self,
...
+        local_dir = os.path.join(self.root, artifact_path)
+        if os.path.exists(local_dir):
+            version = len(os.listdir(local_dir)) + 1
+        else:
+            version = 1
+        artifact_path = os.path.join(artifact_path, str(version))
+        local_dir = os.path.join(local_dir, str(version))
+        os.makedirs(local_dir)
+        if model_type == "pytorch":
+            import submarine.models.pytorch
```
[the remark this answers is not in the document]
Comment (jeff-901) [reply submarine-sdk/pysubmarine/submarine/models/client.py:129]: No. I use the function on 131 line.

File: submarine-sdk/pysubmarine/submarine/models/client.py [hunk 3]
```diff
@@ -101,6 +107,39 @@ def save_model(self,
...
+
+            submarine.models.pytorch.save_model(model, local_dir)
+        elif model_type == "tensorflow":
+            import submarine.models.tensorflow
```
[the remark this answers is not in the document]
Comment (jeff-901) [reply submarine-sdk/pysubmarine/submarine/models/client.py:133]: No. I use the function on 135 line.

File: submarine-sdk/pysubmarine/submarine/models/client.py [hunk 4]
```diff
@@ -101,6 +107,39 @@ def save_model(self,
...
+        pattern = r"[0-9A-Za-z][0-9A-Za-z-_]*[0-9A-Za-z]|[0-9A-Za-z]"
+        if not re.fullmatch(pattern, artifact_path):
+            raise Exception(
+                "Artifact_path must only contains numbers, characters, hyphen and underscore.  \
+            Artifact_path must starts and ends with numbers or characters.")
+        local_dir = os.path.join(self.root, artifact_path)
+        if os.path.exists(local_dir):
+            version = len(os.listdir(local_dir)) + 1
+        else:
+            version = 1
```
[the remark this answers is not in the document]
Comment (jeff-901) [reply submarine-sdk/pysubmarine/submarine/models/client.py:124]: That sounds good. I will edit it.

File: submarine-sdk/pysubmarine/submarine/models/client.py (hunk 1 above)
Comment (jeff-901) [reply submarine-sdk/pysubmarine/submarine/models/client.py:140]: That sounds correct. I will delete all the local model.

**Review by pingsutw — COMMENTED** [review]
Other LGTM

File: submarine-sdk/pysubmarine/submarine/artifacts/Repository.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
```
[the remark this answers is not in the document]
Comment (jeff-901) [reply submarine-sdk/pysubmarine/submarine/artifacts/Repository.py:1]: ok.

File: dev-support/docker-images/submarine/create_bucket.sh [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -euo pipefail
+
+S3_ENDPOINT_URL="http://submarine-minio-service:9000"
+AWS_ACCESS_KEY_ID="submarine_minio"
+AWS_SECRET_ACCESS_KEY="submarine_minio"
+
+/bin/bash -c "sleep 60; mc config host add minio ${S3_ENDPOINT_URL} ${AWS_ACCESS_KEY_ID} ${AWS_SECRET_ACCESS_KEY}"
```
[the remark this answers is not in the document]
Comment (jeff-901) [reply dev-support/docker-images/submarine/create_bucket.sh:27]: This is for waiting the minio pod to setup. I could add a comment.

**Comment by jeff-901** [comment]
@pingsutw  I add the kubectl to wait for the minio pod ready. Repository uses the environmental variables. User can set the environmental variables by providing args when creating the ModelsClient.

**Review by pingsutw — APPROVED** [review]
LGTM

**Comment by pingsutw** [comment]
@jeff-901 Could you fix the conflict? thanks.

**Comment by jeff-901** [comment]
> @jeff-901 Could you fix the conflict? thanks.

Ok. I have fixed the conflict.

**Pull request closed without merging**